File: design-patterns/oberserver-pattern/cpp/OberserverPattern.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <algorithm>

// Forward declaration of Subject class
class Subject;

// Observer interface
class Observer {
public:
    virtual void update(const Subject& subject) = 0;
    virtual ~Observer() {}
};

// Subject class
class Subject {
public:
    void attach(std::shared_ptr<Observer> observer) {
        observers.push_back(observer);
    }

    void detach(std::shared_ptr<Observer> observer) {
        observers.erase(std::remove(observers.begin(), observers.end(), observer), observers.end());
    }

    void notify() const {
        for (const auto& observer : observers) {
            observer->update(*this);
        }
    }

    virtual int getState() const = 0;
    virtual void setState(int state) = 0;

    virtual ~Subject() {}

private:
    std::vector<std::shared_ptr<Observer>> observers;
};

// Concrete subject class
class ConcreteSubject : public Subject {
public:
    int getState() const override {
        return state;
    }

    void setState(int state) override {
        this->state = state;
        notify();
    }

private:
    int state = 0;
};

// Concrete observer class
class ConcreteObserver : public Observer {
public:
    ConcreteObserver(std::string name) : name(name) {}

    void update(const Subject& subject) override {
        std::cout << name << " received update. New state: " << subject.getState() << std::endl;
    }

private:
    std::string name;
};

int main() {
    // Create a concrete subject
    std::shared_ptr<ConcreteSubject> subject = std::make_shared<ConcreteSubject>();

    // Create concrete observers
    std::shared_ptr<ConcreteObserver> observer1 = std::make_shared<ConcreteObserver>("Observer 1");
    std::shared_ptr<ConcreteObserver> observer2 = std::make_shared<ConcreteObserver>("Observer 2");

    // Attach observers to the subject
    subject->attach(observer1);
    subject->attach(observer2);

    // Set the state of the subject
    subject->setState(10);

    // Detach observer1
    subject->detach(observer1);

    // Set the state of the subject again
    subject->setState(20);

    return 0;
}

File: design-patterns/oberserver-pattern/python/oberserver_pattern.py
from typing import List


class Observer:
    def update(self, message: str) -> None:
        pass


class Subject:
    _observers: List[Observer] = []

    def attach(self, observer: Observer) -> None:
        self._observers.append(observer)

    def detach(self, observer: Observer) -> None:
        self._observers.remove(observer)

    def notify(self, message: str) -> None:
        for observer in self._observers:
            observer.update(message)


class ConcreteObserver(Observer):
    def __init__(self, name: str):
        self.name = name

    def update(self, message: str) -> None:
        print(f"{self.name} received message: {message}")


class ConcreteSubject(Subject):
    def do_something(self) -> None:
        # Perform some actions
        print("Subject is doing something...")
        # Notify observers
        self.notify("Something has happened!")


# Client code
if __name__ == "__main__":
    subject = ConcreteSubject()
    observer1 = ConcreteObserver("Observer 1")
    observer2 = ConcreteObserver("Observer 2")

    subject.attach(observer1)
    subject.attach(observer2)

    subject.do_something()

    subject.detach(observer2)

    subject.do_something()
